dpnp/backend/extensions/statistics/histogramdd.cpp

```diff
@@ -0,0 +1,357 @@
+        HistType *out = static_cast<HistType *>(vout);
+
+        auto device = exec_q.get_device();
+
+        const uint32_t local_size = get_max_local_size(exec_q);
+        constexpr uint32_t WorkPI = 128; // empirically found number
+
+        const auto nd_range = make_ndrange(size, local_size, WorkPI);
+
+        return exec_q.submit([&](sycl::handler &cgh) {
+            cgh.depends_on(depends);
+
+            auto dispatch_edges = [&](const uint32_t local_mem,
+                                      const auto &weights, auto &hist) {
+                if (device.is_gpu() && (local_mem >= total_edges)) {
+                    auto edges = CachedEdgesDd<BinsT, EdgesCountT>(
+                        bins_edges, total_edges, bins_edges_count, dims, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+                else {
+                    auto edges = UncachedEdgesDd<BinsT, EdgesCountT>(
+                        bins_edges, total_edges, bins_edges_count, dims, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+            };
+
+            auto dispatch_bins = [&](const auto &weights) {
+                auto local_mem_size = get_local_mem_size_in_items<T>(device);
+                local_mem_size -= 2 * dims; // for min-max values
+                local_mem_size -= CeilDiv(dims * sizeof(EdgesCountT),
+                                          sizeof(T)); // for edges count
+
+                if (local_mem_size >= bins_count) {
+                    const auto local_hist_count = get_local_hist_copies_count(
+                        local_mem_size, local_size, bins_count);
+
+                    auto hist = HistWithLocalCopies<HistType>(
+                        out, bins_count, local_hist_count, cgh);
+                    const uint32_t free_local_mem =
+                        local_mem_size - hist.size();
+
+                    dispatch_edges(free_local_mem, weights, hist);
+                }
+                else {
+                    auto hist = HistGlobalMemory<HistType>(out);
+                    auto edges = UncachedEdgesDd<BinsT, EdgesCountT>(
+                        bins_edges, total_edges, bins_edges_count, dims, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+            };
+
+            if (weights) {
+                auto _weights = Weights(weights);
+                dispatch_bins(_weights);
+            }
+            else {
+                auto _weights = NoWeights();
+                dispatch_bins(_weights);
+            }
+        });
+    }
+};
+
+template <typename T, typename HistType = size_t>
+using HistogramddF2 = HistogramddF<T, T, HistType>;
```

Comment:
following implementation in `histogram.cpp`:
```suggestion
using HistogramddF_ = HistogramddF<T, T, HistType>;
```